openstf/pipeline/train_predict_backtest.py

```diff
@@ -0,0 +1,135 @@
+            These horizons are also used to make predictions (one for every horizon)
+
+    Returns:
+        forecast (pandas.DataFrame)
+
+    """
+    if training_horizons is None:
+        training_horizons = DEFAULT_TRAIN_HORIZONS
+
+    # Validate and clean data
+    validated_data = validation.clean(validation.validate(input_data))
+
+    # Check if sufficient data is left after cleaning
+    if not validation.is_data_sufficient(validated_data):
+        raise ValueError(
+            f"Input data is insufficient for {pj['name']} "
+            f"after validation and cleaning"
+        )
+
+    # Add features
+    data_with_features = TrainFeatureApplicator(
+        horizons=training_horizons, feature_names=pj["feature_names"]
+    ).add_features(validated_data)
+
+    # Split data
+    train_data, validation_data, test_data = split_data_train_validation_test(
+        data_with_features.sort_index(axis=0), test_fraction=0.15, back_test=True
+    )
+
+    # Create relevant model
+    model = ModelCreator.create_model(pj)
+
+    # split x and y data
+    train_x, train_y = train_data.iloc[:, 1:-1], train_data.iloc[:, 0]
+    validation_x, validation_y = (
+        validation_data.iloc[:, 1:-1],
+        validation_data.iloc[:, 0],
+    )
+
+    # Configure evals for early stopping
+    eval_set = [(train_x, train_y), (validation_x, validation_y)]
+
+    model.set_params(**pj["hyper_params"])
+    model.fit(
+        train_x,
+        train_y,
+        eval_set=eval_set,
+        early_stopping_rounds=DEFAULT_EARLY_STOPPING_ROUNDS,
+        verbose=False,
+    )
+    logging.info("Fitted a new model, not yet stored")
+
+    # Do confidence interval determination
+    model = StandardDeviationGenerator(
+        pj, validation_data
+    ).generate_standard_deviation_data(model)
+
+    # Report about the training procces
+    report = Reporter(pj, train_data, validation_data, test_data).generate_report(model)
+
+    # Save model
+    serializer = PersistentStorageSerializer(trained_models_folder)
+    serializer.save_model(model, pid=pj["id"])
+
+    # Save reports/figures
+    report.save_figures(save_path=save_figures_folder)
+
+    # Load most recent model for the given pid
+    model = PersistentStorageSerializer(
+        trained_models_folder=trained_models_folder
+    ).load_model(pid=pj["id"])
+
+    # Predict
+    model_forecast = model.predict(test_data.iloc[:, 1:-1])
+    forecast = pd.DataFrame(index=test_data.index, data={"forecast": model_forecast})
+
+    # Add confidence
+    forecast = ConfidenceIntervalApplicator(
+        model, test_data.iloc[:, 1:-1]
+    ).add_confidence_interval(forecast, pj)
+
+    # Prepare for output
+    forecast = add_prediction_job_properties_to_forecast(
+        pj,
+        forecast,
+    )
+
+    # Add column with realised load and horizon information
+    forecast["realised"] = test_data.iloc[:, 0]
+    forecast["Horizon"] = test_data.iloc[:, -1]
+
+    return forecast
```

Comment:
Should we also include some method for evaluation of the backtest. E.g. a report on the performance of the model?
Can be very lightweight.

I think you can reuse some of the plot functions from https://github.com/Alliander/icarus-ops-gui/blob/master/app/pages/predictability.py and maybe also from the already available 'Reporter'

Reply:
I have added the evaluation of the backtest to the notebook using the openstf.metrics metrics